### PR TITLE
Adding event queries to db

### PR DESCRIPTION
**Pull request by slundqui**
This PR is the first of a series of PRs to support multi-pool trading in agent0.

In this PR, we move away from exposing the `agent.wallet` object (as we're deprecating doing bookkeeping on the wallet itself in python) in favor of using the `agent.get_positions()` function. This function (1) does a query of the chain to gather events and adds them to a `TradeEvent` db table, and (2) queries from the `TradeEvent` table to get the current positions a wallet has.

The `TradeEvent` table handles all events on any hyperdrive tokens (i.e., long/short/lp). There's a bit of overlap with the `WalletDelta` table, with the main exception that the `TradeEvent` table is lazy - the table only gets updated when `agent.get_positions()` gets called, and only with the events from `agent`. In addition, the table handles both trade events (e.g., `OpenLong`) and single transfer trades (e.g., wallet to wallet transfers of tokens). We likely can deprecate the `WalletDelta` table with a special call to gather all trade events from a Hyperdrive pool, which fills the `TradeEvent` table with every wallet that has made a trade on the pool.

There are a couple of places that can be optimized. Currently, we query the chain for events for every `get_positions` call (from the latest entry in the db to latest block). Some bookkeeping is needed to e.g., don't get events from the logs if a user calls `get_positions` on the same block.

As a temporary fix, we also remove `agent.wallet` from remote chains, and `get_positions` gathers all events from the remote chain each time it's called. This will get fixed once the database is exposed in the underlying chain object, with the remote wallet also using the `TradeEvent` table to gather wallet positions.

Final note: the failing test here is fixed in https://github.com/delvtech/agent0/pull/1462.

## Changes
- Removing interactive wallet in favor of using `get_positions()`.
- Added `TradeEvent` table to database, with supporting ingestion (`trade_events_to_db`) and query (`get_positions_from_db` and `get_trade_events`) interface functions.
- Changed all interactive `agent.wallet` calls to `agent.get_positions()`
- Renamed `contract_address` to `hyperdrive_address` in `PoolConfig` db table.

### CODE REVIEW

File: src/agent0/chainsync/db/hyperdrive/chain_to_db.py [hunk 1]
```diff
@@ -100,3 +113,325 @@ def data_chain_to_db(interface: HyperdriveReadInterface, block: BlockData, sessi
...
+
+    # Get the earliest block to get events from
+    # TODO can narrow this down to the last block we checked
+    # For now, keep this as the latest entry of this wallet.
+    # + 1 since the queries are inclusive
+    from_block = get_latest_block_number_from_trade_event(db_session, wallet_addr) + 1
+
+    # Gather all events we care about here
+    # Transfers to and from wallet address
+    # db_event_objs: dict[str, HyperdriveEvent] = {}
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/chain_to_db.py:161]: > `    # db_event_objs: dict[str, HyperdriveEvent] = {}`

 dead code? delete it if so

File: src/agent0/chainsync/db/hyperdrive/chain_to_db.py [hunk 2]
```diff
@@ -100,3 +113,325 @@ def data_chain_to_db(interface: HyperdriveReadInterface, block: BlockData, sessi
...
+    interfaces: list[HyperdriveReadInterface],
+    wallet_addr: str,
+    db_session: Session,
+) -> None:
+    """Function to query trade events from all pools and add them to the db.
+
+    Arguments
+    ---------
+    interfaces: list[HyperdriveReadInterface]
+        The hyperdrive interface objects connected to a pool.
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/chain_to_db.py:144]: It didn't click at first that we associate an interface with a specific pool. Makes sense now, but maybe updating this docstring to emphasize that **each** object is connected to a pool would help people get it more quickly:

```
A collection of Hyperdrive interface objects, each connected to a pool.
```

File: src/agent0/chainsync/db/hyperdrive/chain_to_db.py [hunk 3]
```diff
@@ -100,3 +113,325 @@ def data_chain_to_db(interface: HyperdriveReadInterface, block: BlockData, sessi
...
+        )
+        all_events.extend([_event_data_to_dict(event) for event in events])
+
+    # Convert to dataframe
+    events_df = pd.DataFrame(all_events)
+    # If no events, we just return
+    if len(events_df) == 0:
+        return
+
+    # Each transaction made through hyperdrive has two rows for each transaction,
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/chain_to_db.py:228]: "each transaction" used twice; delete second one?

File: src/agent0/chainsync/db/hyperdrive/chain_to_db.py [hunk 4]
```diff
@@ -100,3 +113,325 @@ def data_chain_to_db(interface: HyperdriveReadInterface, block: BlockData, sessi
...
+            lambda x: Decimal(x) / Decimal(1e18)  # type: ignore
+        )
+        # Base delta is always 0
+        transfer_events_df["base_delta"] = Decimal(0)
+
+    # Drop all transfer single events
+    events_df = events_df[events_df["event"] != "TransferSingle"].reset_index(drop=True)
+
+    # Sanity check, one hyperdrive event per transaction hash
+    assert events_df.groupby("transactionHash")["event"].nunique().all() == 1
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/chain_to_db.py:294]: better to put this in a conditional and then raise an error

- works well with down-stream error handling
- running the program in optimized mode (by setting the `PYTHONOPTIMIZE` command line flag, `-O`) silently disables assert statements.

File: src/agent0/chainsync/db/hyperdrive/chain_to_db.py [hunk 5]
```diff
@@ -100,3 +113,325 @@ def data_chain_to_db(interface: HyperdriveReadInterface, block: BlockData, sessi
...
+    out["args"] = dict(in_val["args"])
+
+    # Convert transaction hash to string
+    out["transactionHash"] = in_val["transactionHash"].to_0x_hex()
+    # Get token id field from args.
+    # This field is `assetId` for open/close long/short
+    return out
+
+
+# TODO cleanup
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/chain_to_db.py:130]: This function is quite a beast, and has some mission-critical components. I think we should prioritize cleanup **and** testing as soon as this refactor is done.

File: src/agent0/chainsync/db/hyperdrive/interface.py [hunk 6]
```diff
@@ -19,10 +19,114 @@
...
     WalletDelta,
     WalletPNL,
 )
 
+# Event Data Ingestion Interface
+
+
+def add_transfer_events(transfer_events: list[TradeEvent], session: Session) -> None:
+    """Add transfer events to the transfer events table.
+    Arguments
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/interface.py:32]: newline before Arguments

File: src/agent0/chainsync/db/hyperdrive/interface.py [hunk 7]
```diff
@@ -19,10 +19,114 @@
...
+        session.commit()
+    except exc.DataError as err:
+        session.rollback()
+        logging.error("Error adding transaction: %s", err)
+        raise err
+
+
+def get_latest_block_number_from_trade_event(session: Session, wallet_addr: str) -> int:
+    """Get the latest block number based on the hyperdrive events table in the db.
+    Arguments
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/interface.py:51]: Newline before Arguments

File: src/agent0/chainsync/db/hyperdrive/interface.py [hunk 8]
```diff
@@ -19,10 +19,114 @@
...
+    """
+
+    query = session.query(func.max(TradeEvent.block_number)).filter(TradeEvent.wallet_address == wallet_addr).scalar()
+    if query is None:
+        return 0
+    return int(query)
+
+
+def get_trade_events(session: Session, wallet_addr: str | None = None) -> pd.DataFrame:
+    """Get all trade events and returns as a pandas dataframe.
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/interface.py:71]: "...return a Pandas dataframe."

File: src/agent0/chainsync/db/hyperdrive/interface.py [hunk 9]
```diff
@@ -19,10 +19,114 @@
...
+    ---------
+    session: Session
+        The initialized db session object.
+    wallet_addr: str | None, optional
+        The wallet address to filter the results on. Return all if None
+
+    Returns
+    -------
+    DataFrame
+        A DataFrame that consists of the queried trade events data
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/interface.py:83]: period at the end

File: src/agent0/chainsync/db/hyperdrive/interface.py [hunk 10]
```diff
@@ -19,10 +19,114 @@
...
+
+def get_trade_events(session: Session, wallet_addr: str | None = None) -> pd.DataFrame:
+    """Get all trade events and returns as a pandas dataframe.
+
+    Arguments
+    ---------
+    session: Session
+        The initialized db session object.
+    wallet_addr: str | None, optional
+        The wallet address to filter the results on. Return all if None
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/interface.py:78]: period at the end

File: src/agent0/chainsync/db/hyperdrive/interface.py [hunk 11]
```diff
@@ -19,10 +19,114 @@
...
+    """Gets all positions for a given wallet address.
+
+    Arguments
+    ---------
+    session: Session
+        The initialized db session object.
+    wallet_addr: str
+        The wallet address to filter the results on.
+    hyperdrive_address: str
+        The hyperdrive address to filter the results on.
```
Comment (dpaiton) [inline src/agent0/chainsync/db/hyperdrive/interface.py:101]: , optional

File: src/agent0/chainsync/df_to_db.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+MAX_BATCH_SIZE = 10000
+
+
+def df_to_db(insert_df: pd.DataFrame, schema_obj: Type[Base], session: Session):
+    """Helper function to add a dataframe to a database.
+
+    Arguments
+    ---------
+    insert_df: pd.DataFrame
+        The dataframe to insert
```
Comment (dpaiton) [inline src/agent0/chainsync/df_to_db.py:21]: period at the end of arg descriptors

**Review by dpaiton — APPROVED** [review]
Looks great w/ minor docstring nits.

File: src/agent0/chainsync/db/hyperdrive/chain_to_db.py [hunk 13]
```diff
@@ -100,3 +113,325 @@ def data_chain_to_db(interface: HyperdriveReadInterface, block: BlockData, sessi
...
+            fromBlock=from_block,
+            argument_filters={"provider": wallet_addr},
+        )
+        all_events.extend([_event_data_to_dict(event) for event in events])
+
+    # Convert to dataframe
+    events_df = pd.DataFrame(all_events)
+    # If no events, we just return
+    if len(events_df) == 0:
+        return
```
[the remark this answers is not in the document]
Comment (slundqui) [reply src/agent0/chainsync/db/hyperdrive/chain_to_db.py:226]: This actually throws a lint error. Overall, this function shouldn't return anything at all (as opposed to returning None). Keep this as is.

**Pull request merged**